DEVELOPMENT.md

```diff
@@ -0,0 +1,70 @@
+# Development documentation
+
+This document describes how to set up the development environment and how to develop Vald.
+
+## Set up environment
+
+### Prerequisites
+
+#### OS
+
+- When using Docker related environment, you can use any OS that supports Docker.
+- When using native environment, `Linux` is required.
+
+#### Architecture
+
+`amd64` is recommended because `NGT`, the vector search library we use, is optimized for `amd64`.
+But you can also build and test `Vald` on `arm64` with the same way as described below.
+
+### Devcontainer
+
+This is the easiest way to start developing `Vald`. You can just open our [devcontainer.json](./.devcontainer/devcontainer.json) with `VS Code` and go.
+
+### Other
+
+We don't officially have a setup documentation for now, but you can take a look at the [`Dockerfile`](./dockers/dev/Dockerfile).
+That's everything you need to build and test `Vald`, so you can use it as a reference.
+
+> If you would like to use the `Dockerfile` directlly, please note that `docker-in-docker` environment is required to run our E2E tests.
+> In devcontainer, [`VS Code` handles it for us](https://github.com/devcontainers/features/tree/main/src/docker-in-docker).
+
+## Run tests
+
+### Unit tests
+
+The command below will run all the unit tests.
+
+```bash
+make test
+```
+
+### E2E tests
+
+The steps below will deploy `Vald` to local `k3d` cluster and run the E2E tests.
+
+1. Change `example/helm/values.yaml` to `dimensions: 784` and `distance_type: l2`.
+2. Run the commands below.
+
+```bash
+# Download the dataset
+make hack/benchmark/assets/dataset/fashion-mnist-784-euclidean.hdf5
+
+# Create k3d cluster
+k3d cluster create vald
```

Comment:
How about the following? 🤔 :
```suggestion
make k3d/start
```

Reply:
`make k3d/start` creates a lot of pods that are not required for running the E2E test, leading to a log wait until those pods become ready. So I think just creating a plain k3d is better here.